src/ElasticApmAgentStartupHook/StartupHook.cs

```diff
@@ -104,7 +108,53 @@ public static void Initialize()
 			}
 		}
 
-		InvokerLoaderMethod(loader);
+		if (loaderAssembly is null)
+		{
+			_logger.WriteLine(
+				$"No {LoaderDll} assembly loaded. Agent not loaded");
+		}
+
+		LoadAssembliesFromLoaderDirectory(loaderDirectory);
+		InvokerLoaderMethod(loaderAssembly);
+	}
+
+	/// <summary>
+	/// Loads assemblies from the loader directory if they exist
+	/// </summary>
+	/// <param name="loaderDirectory"></param>
+	private static void LoadAssembliesFromLoaderDirectory(string loaderDirectory)
+	{
+		var context = new ElasticApmAssemblyLoadContext();
+		AssemblyLoadContext.Default.Resolving += (_, name) =>
+		{
+			var assemblyPath = Path.Combine(loaderDirectory, name.Name + ".dll");
+			if (File.Exists(assemblyPath))
+			{
+				try
+				{
+					var assemblyName = AssemblyName.GetAssemblyName(assemblyPath);
+					if (name.Version == assemblyName.Version)
+					{
+						var keyToken = name.GetPublicKeyToken();
+						var assemblyKeyToken = assemblyName.GetPublicKeyToken();
+						if (keyToken.SequenceEqual(assemblyKeyToken))
+						{
+							// load Elastic.Apm assemblies with the default assembly load context, to allow DiagnosticListeners to subscribe.
+							// For all other dependencies, load with a separate load context to not conflict with application dependencies.
+							return name.Name.StartsWith("Elastic.Apm", StringComparison.Ordinal)
```

Comment:
Are we totally sure about this? 🤔 

This means if we have let's say `Elstic.Apm.X` which depends on assembly `Y`, we load `Elastic.Apm.X`  into the default context, but load `Y` into the custom context - in this case `Elastic.Apm.X` would have missing typerefs in the default load context.

This seems to run and with our current references it does not seem to have any problems, but this looks a bit strange to me.

Reply:
There are two problems that this is attempting to address:

1. The Elastic.Apm assemblies that contain Diagnostic Listeners must be loaded into the default load context to be able to subscribe to `DiagnosticListeners.AllListeners` and observe diagnostic events - subscription will not occur if loaded into a different load context.
2. Dependencies of Elastic.Apm assemblies cannot version conflict with dependencies of the application. In testing and debugging, the dependencies of the internalized Ben.Demystifier, namely, System.Reflection.Metadata, can very likely be a different version to the one referenced by the application, which was true for ASP.NET Core 3.0 if I recall.

Are there additional tests that can be written to provide more assurance in this approach?

Reply:
> Are there additional tests that can be written to provide more assurance in this approach?

I did some manual testing and so far everything worked. Maybe later we could add more tests where:
- We make sure we also trigger stack trace capturing (e.g. making an outgoing HTTP call in the sample app)
- Where we have agent assemblies with additional dependencies (e.g. using elastisearch-net and make the app depend on a different version than the agent)

I think those will trigger code paths that are critical. I did those tests manually and those worked for me.